src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1104,15 +1104,33 @@ bool ObjectSynchronizer::is_async_deflation_needed() {
     // Async deflation request.
     return true;
   }
+
+  jlong time_since_last = time_since_last_async_deflation_ms();
+
   if (AsyncDeflationInterval > 0 &&
-      time_since_last_async_deflation_ms() > AsyncDeflationInterval &&
+      time_since_last > AsyncDeflationInterval &&
       monitors_used_above_threshold(&_in_use_list)) {
     // It's been longer than our specified deflate interval and there
     // are too many monitors in use. We don't deflate more frequently
     // than AsyncDeflationInterval (unless is_async_deflation_requested)
     // in order to not swamp the MonitorDeflationThread.
+    log_info(monitorinflation)("Async deflation needed: monitors used are above the threshold");
     return true;
   }
+
+  if (GuaranteedAsyncDeflationInterval > 0 &&
+      time_since_last > GuaranteedAsyncDeflationInterval) {
+    // It's been longer than our specified guaranteed deflate interval.
+    // We need to clean up the used monitors even if the threshold is
+    // not reached, to keep the memory utilization at bay when many threads
+    // touched many monitors. This inflation is expected to have no progress
+    // in normal conditions, so it should not count towards non-progress
+    // adjustments.
+    _no_progress_cnt = 0;
```

Comment:
Not sure about this one? I understand your intention, but I think setting `_no_progress_cnt` to zero doesn't really improve things. If there will be deflations, `_no_progress_cnt` will be set to zero anyway (in `deflate_idle_monitors()`. However if there are no deflations, `_no_progress_cnt` will now always become `1`, even if it was higher before.

So maybe doing `--_no_progress_cnt` here would be more accurate?

Reply:
We don't know if this guaranteed deflation would make progress or not, so we cannot decrement blindly. In the new commit, I just introduced a new flag that disallows progress updates when we are in guaranteed deflation path.

Reply:
The problem is here: https://github.com/openjdk/jdk/blob/1fd400608e9ea423a6e4f1797652bd02f027da1b/src/hotspot/share/runtime/synchronizer.cpp#L1072-L1073

Along with the reset of `_no_progress_cnt` in `monitors_used_above_threshold`, we *also* bump the ceiling, which means the next checks for used monitor thresholds get more and more unlikely to fire, because ceiling goes farther and farther away. You can clearly see that in the `-Xlog:monitordeflation` without the remediation like this: with every `NoAsyncDeflationProgressMax` guaranteed attempts, that block bumps the ceiling. :)

Reply:
I agree, I just don't understand why `--_no_progress_cnt` wouldn't have the same effect like your current solution? With `--_no_progress_cnt` in the block for the `GuaranteedAsyncDeflationInterval` case in `is_async_deflation_needed()` a call to `deflate_idle_monitors()` wouldn't increment the initial value of `_no_progress_cnt` which is exactly the same behavior you have now with the introduction of new `_no_progress_allow_updates` parameter.

Reply:
Ah, I understand. Yes, decrement would work too. But I think the explicit flag is cleaner, if only it took me three comments to understand how would that work :)

Reply:
Ok, glad we agree on that one now :)

However, I just realized that with the explicit flag you now do not reset `_no_progress_cnt` back to zero in the case where you do deflate monitors in a round triggered by `GuaranteedAsyncDeflationInterval`. Wouldn't it be better to do that?

Reply:
No, I don't think so, mostly because this patch tries to avoid interactions with any other triggering code, so that it does not break it accidentally. Let's isolate the triggers instead of trying to make sure they work together well.

Consider a case where there is a sequence of several no-progress threshold cleanups, this would normally collect enough `_no_progress_cnt`-s to bump the threshold ceiling in order to break the sequence. Note that no-progress means we tried to deflate, but deflated none. So if there is a guaranteed cleanup in the middle of that sequence which drops `_no_progress_cnt = 0` and still does not deflate anything (which is likely), the threshold cleanup code would start another series of cleanups from the beginning. And this is the simplest interaction example I can think of...

Reply:
Sorry for being so impertinent :) but I think "*if there is a guaranteed cleanup in the middle of that sequence which drops `_no_progress_cnt` = 0 and still does not deflate anything*" can not happen. `_no_progress_cnt` will only be dropped to zero if there was progress (i.e. if we really did deflate).

I think the problem is exactly the other way round. Imagine we do a sequence of several no-progress threshold cleanups, which collect some `_no_progress_cnt`-s but not enough to bump the threshold ceiling in order to break the sequence because eventually one of the threshold cleanups deflates some monitors and resets `_no_progress_cnt` to zero. Now with your change this cleanup which deflated some monitors might happen during a guaranteed cleanup and because of your new flag it will not reset `_no_progress_cnt` as before and a few follow up no-progress threshold cleanups will then unnecessarily bump the threshold ceiling.

So I still think the `--_no_progress_cnt` would be the more "original behavior-preserving" solution, or if you want to go with the extra flag then probably something like:
```cpp
    if (deflated_count != 0) {
      _no_progress_cnt = 0;
    } else {
      if (_no_progress_allow_updates) {
        _no_progress_cnt++;
      }
    }
```

Reply:
Okay, that is a solid argument. See new commit :)